added json encoding to ajax

cscortes opened the pull request:
Per your request, added the data-encoding as kt-data-encoding.  Due to all the cool new changes to kutty, I had to do minimal additions to the code. 

```
           <form kt-post="http://localhost:9099/auth" kt-data-encoding="json" >

                <div class="form-group">
                    <label>Username</label>
                    <input type="text" class="form-control" name="username">
                </div>
            
                <div class="form-group">
                    <label>Password</label>
                    <input type="password" class="form-control" name="password">
                </div>
            
            <button class="btn btn-default">Submit</button>
            </form>         
```

chg20 commented:
Thank you!  I want to spend some time thinking about this because I want to minimize the number of attributes on kutty.  Maybe we can make the encoding an option on the request declaration?

```html
<div kt-get="https://foo.bar encoding:json">Click Me</div>
```

As well as a configurable option:

```javascript
  kutty.config.defaultEncoding = "json"
```
?  

I'm settling in on the white-space separated arguments with colons pattern for other attributes too (e.g. kt-swap="innerHTML swap:1s settle:1s") and I'm wondering if we should use that to keep the number of attributes down for most people.

Let's think about it for a bit...

zodman commented:
i suggest like alpinejs do: kt-get.json_encode=""

chg20 commented:
“You best start believing in custom attribute modifiers @zodman... you’re creating one.”

https://m.youtube.com/watch?v=1fzXmJyolfY

cscortes commented:
Added feature requested, default value for encoding.  Also, it can handle URL with encoding: json like so:

```
       <form kt-post="http://localhost:9099/auth encoding:json" >

                <div class="form-group">
                    <label>Username</label>
                    <input type="text" class="form-control" name="username">
                </div>
            
                <div class="form-group">
                    <label>Password</label>
                    <input type="password" class="form-control" name="password">
                </div>
            
            <button class="btn btn-default">Submit</button>
```
Still need some tests to make sure I didn't break other code.  

NOTE:
also adding fix for scary errors on chrome.  

NOTE:
this new implementation also removes the need for path variable in:

function issueAjaxRequest(elt, verb, path, eventTarget)

cscortes commented:
per request here is my patch file.  HTMX came in between my changes, so this is a patch from the kutty base.  

```
diff --git a/src/kutty.js b/src/kutty.js
index 35ce315..2b71f7f 100644
--- a/src/kutty.js
+++ b/src/kutty.js
@@ -1084,6 +1084,31 @@ var kutty = kutty || (function () {
             return swapSpec;
         }
 
+        function getUrlSpecification(elt, verb) {
+            // verb should be kt-post, kt-get, kt-put, etc
+            var ktverb = "kt-" + verb;
+            var info = getClosestAttributeValue(elt, ktverb);
+            var urlSpec = {
+                "encoding": kutty.config.defaultEncoding,
+                "url": null,
+            }
+
+            if (info) {
+                var split = splitOnWhitespace(info);
+                if (split.length > 0) {
+                    urlSpec["url"] = split[0];
+                    for (var i = 1; i < split.length; i++) {
+                        var modifier = split[i];
+                        if ((modifier.indexOf("encoding:") === 0) && (modifier.indexOf("json") >= 0)) {
+                            urlSpec["encoding"] = "json";
+                        }
+                    }
+                }
+            }
+            return urlSpec;
+        }
+
+
         function issueAjaxRequest(elt, verb, path, eventTarget) {
             var target = getTarget(elt);
             if (target == null) {
@@ -1115,12 +1140,17 @@ var kutty = kutty || (function () {
             var headers = getHeaders(elt, target, prompt, eventTarget);
             var rawParameters = getInputValues(elt, verb);
             var filteredParameters = filterValues(rawParameters, elt, verb);
+            var urlSpec = getUrlSpecification(elt, verb);
 
             if (verb !== 'get') {
                 headers['Content-Type'] = 'application/x-www-form-urlencoded; charset=UTF-8';
                 if (verb !== 'post') {
                     headers['X-HTTP-Method-Override'] = verb.toUpperCase();
                 }
+
+                if (urlSpec.encoding == "json") {
+                    headers['Content-Type'] = 'application/json';
+                }
             }
 
             var requestConfig = {
@@ -1136,14 +1166,16 @@ var kutty = kutty || (function () {
             var requestURL;
             if (verb === 'get') {
                 var noValues = Object.keys(filteredParameters).length === 0;
-                requestURL = path + (noValues ? "" : "?" + urlEncode(filteredParameters));
+                requestURL = urlSpec.url + (noValues ? "" : "?" + urlEncode(filteredParameters));
                 xhr.open('GET', requestURL, true);
             } else {
-                requestURL = path;
+                requestURL = urlSpec.url;
                 xhr.open('POST', requestURL, true);
             }
 
-            xhr.overrideMimeType("text/html");
+            if (urlSpec.encoding != "json") {
+                xhr.overrideMimeType("text/html");
+            }
 
             // request headers
             for (var header in headers) {
@@ -1157,8 +1189,12 @@ var kutty = kutty || (function () {
                 try {
                     if (!triggerEvent(elt, 'beforeOnLoad.kutty', eventDetail)) return;
 
-                    handleTrigger(elt, this.getResponseHeader("X-KT-Trigger"));
-                    var pushedUrl = this.getResponseHeader("X-KT-Push");
+                    var ktTrigger =  (this.getAllResponseHeaders().indexOf("X-KT-Trigger") >= 0 ) ?
+                    this.getResponseHeader("X-KT-Trigger") : null;
+                    handleTrigger(elt, ktTrigger );
+
+                    var pushedUrl = (this.getAllResponseHeaders().indexOf("X-KT-Push") >= 0 ) ? 
+                    this.getResponseHeader("X-KT-Push") : null ;
 
                     var shouldSaveHistory = shouldPush(elt) || pushedUrl;
 
@@ -1236,7 +1272,12 @@ var kutty = kutty || (function () {
             }
             if(!triggerEvent(elt, 'beforeRequest.kutty', eventDetail)) return endRequestLock();
             addRequestIndicatorClasses(elt);
-            xhr.send(verb === 'get' ? null : urlEncode(filteredParameters));
+           
+            if (urlSpec.encoding == "json") {
+                xhr.send(JSON.stringify(filteredParameters));
+            } else {
+                xhr.send(verb === 'get' ? null : urlEncode(filteredParameters));
+            }
         }
 
         //====================================================================

```

cscortes commented:
Cool, I sync with my upstream (once I set it :) and now I am synced with my branch and htmx.  Also, because of existing test cases I noticed that my additions broke "boost" and so I fixed that and verified that the test cases passed again.  And finally, I added some new test cases of my own for verb json encoding.

cscortes commented:
Added defaultEncoding variable and value "xml" to htmx.config, opps.  Tests failed, so need to make a minor change to the testcases, now all is well again.

chg20 commented on src/htmx.js:
```diff
@@ -1319,7 +1361,8 @@ var htmx = htmx || (function () {
                 defaultSettleDelay:100,
-                includeIndicatorStyles:true
+                includeIndicatorStyles:true,
+                defaultEncoding: "xml"
```
I think the default should be `x-www-form-urlencoded`, not XML (it isn't XML format)

cscortes replied:
yes, that makes sense.  Still do you really want people to type all that out.  how about:
defaultEncoding = "url";
the only other option being:
defaultEncoding = "json";
NOTE:
The code only looks for "json" anything else is automatically considered:  x-www-form-urlencoded

cscortes replied:
Per your okay, we will go with "url".

chg20 commented on test/core/verbencoding.js:
```diff
@@ -0,0 +1,134 @@
+    it('handles post with form parameters', function () {
+
+        this.server.respondWith("POST", "/test", function (xhr) {
+            var values = JSON.parse(xhr.requestBody);
```
I think it is worth testing the parameter values here as well

cscortes replied:
ok, will add it.

cscortes replied:
Added: 
```
            values["username"].should.be.equal("joe");
            values["password"].should.be.equal("123456");
```
to 5 tests.

chg20 commented on test/core/verbencoding.js:
```diff
@@ -0,0 +1,134 @@
+
+        byId("btnSubmit").click();
+        this.server.respond();
+    })
```
And maybe add a response value based on the submitted values to test here to confirm the round trip?

cscortes replied:
Added:

```
            var ans = { "passwordok": values["password"] == "123456"};
            xhr.respond(200, {}, JSON.stringify(ans));
```

and

```
     this.server.lastRequest.response.should.equal('{"passwordok":true}');
```
to 5 tests.

cscortes commented:
Now using the new extensions tool.  Way cool results. I had too much stuff with my own fork, so I I am including a patch against "dev" for the changes.  


```
diff --git a/.gitignore b/.gitignore
index b667ce0..bf082e6 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,5 @@
 .idea
 /node_modules
 _site
-test/scratch.html
\ No newline at end of file
+test/scratch.html
+patch.sh
diff --git a/src/htmx.js b/src/htmx.js
index 250b208..9a62ceb 100644
--- a/src/htmx.js
+++ b/src/htmx.js
@@ -1116,9 +1116,13 @@ var htmx = htmx || (function () {
             var encodedParameters = null;
             forEach(getExtensions(elt), function (extension) {
                 if (encodedParameters == null) {
-                    extension.encodeParameters(xhr, filteredParameters, elt);
+                    encodedParameters = extension.encodeParameters(xhr, filteredParameters, elt);
                 }
             });
+            
+            if (encodedParameters != null) {
+                return encodedParameters;
+            }
             return urlEncode(filteredParameters);
         }
 
@@ -1209,8 +1213,12 @@ var htmx = htmx || (function () {
                 try {
                     if (!triggerEvent(elt, 'beforeOnLoad.htmx', eventDetail)) return;
 
-                    handleTrigger(elt, this.getResponseHeader("X-HX-Trigger"));
-                    var pushedUrl = this.getResponseHeader("X-HX-Push");
+                    var hxTrigger =  (this.getAllResponseHeaders().indexOf("X-HX-Trigger") >= 0 ) ?
+                    this.getResponseHeader("X-HX-Trigger") : null;
+                    handleTrigger(elt, hxTrigger );
+
+                    var pushedUrl = (this.getAllResponseHeaders().indexOf("X-HX-Push") >= 0 ) ?
+                    this.getResponseHeader("X-HX-Push") : null ;
 
                     var shouldSaveHistory = shouldPush(elt) || pushedUrl;
 
diff --git a/test/extensions/json-enc.js b/test/extensions/json-enc.js
new file mode 100644
index 0000000..d2efefb
--- /dev/null
+++ b/test/extensions/json-enc.js
@@ -0,0 +1,143 @@
+// 
+describe("htmx AJAX Verbs with JSON encoding", function() {
+    beforeEach(function () {
+        this.server = makeServer();
+        clearWorkArea();
+        htmx.defineExtension('json-enc', {
+            encodeParameters : function(xhr, parameters, elt) {
+                xhr.requestHeaders['Content-Type'] = 'application/json';
+                xhr.overrideMimeType('text/json');
+                return (JSON.stringify(parameters));
+            }
+        });        
+    });
+    afterEach(function () {
+        this.server.restore();
+        clearWorkArea();
+    });
+
+    it('handles basic post properly', function () {
+        var jsonResponseBody = JSON.stringify({});
+        this.server.respondWith("POST", "/test", jsonResponseBody);
+        var div = make("<div hx-post='/test' hx-ext='json-enc'>click me</div>");
+        div.click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal("{}");
+    })
+
+    it('handles basic put properly', function () {
+        var jsonResponseBody = JSON.stringify({});
+        this.server.respondWith("PUT", "/test", jsonResponseBody);
+        var div = make('<div hx-put="/test" hx-ext="json-enc">click me</div>');
+        div.click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal("{}");
+    })
+
+    it('handles basic patch properly', function () {
+        var jsonResponseBody = JSON.stringify({});
+        this.server.respondWith("PATCH", "/test", jsonResponseBody);
+        var div = make('<div hx-patch="/test" hx-ext="json-enc">click me</div>');
+        div.click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal("{}");
+    })
+
+    it('handles basic delete properly', function () {
+        var jsonResponseBody = JSON.stringify({});
+        this.server.respondWith("DELETE", "/test", jsonResponseBody);
+        var div = make('<div hx-delete="/test" hx-ext="json-enc">click me</div>');
+        div.click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal("{}");
+    })
+
+    it('handles post with form parameters', function () {
+
+        this.server.respondWith("POST", "/test", function (xhr) {
+            var values = JSON.parse(xhr.requestBody);
+            values.should.have.keys("username","password");
+            values["username"].should.be.equal("joe");
+            values["password"].should.be.equal("123456");
+            var ans = { "passwordok": values["password"] == "123456"};
+            xhr.respond(200, {}, JSON.stringify(ans));
+        });
+
+        var html = make('<form hx-post="/test" hx-ext="json-enc" > ' + 
+            '<input type="text"  name="username" value="joe"> ' +
+            '<input type="password"  name="password" value="123456"> ' +
+        '<button  id="btnSubmit">Submit</button> ');
+
+        byId("btnSubmit").click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal('{"passwordok":true}');
+    })
+
+    it('handles put with form parameters', function () {
+
+        this.server.respondWith("PUT", "/test", function (xhr) {
+            var values = JSON.parse(xhr.requestBody);
+            values.should.have.keys("username","password");
+            values["username"].should.be.equal("joe");
+            values["password"].should.be.equal("123456");
+            var ans = { "passwordok": values["password"] == "123456"};
+            xhr.respond(200, {}, JSON.stringify(ans));
+        });
+
+        var html = make('<form hx-put="/test" hx-ext="json-enc" > ' + 
+            '<input type="text"  name="username" value="joe"> ' +
+            '<input type="password"  name="password" value="123456"> ' +
+        '<button  id="btnSubmit">Submit</button> ');
+
+        byId("btnSubmit").click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal('{"passwordok":true}');
+    })
+
+
+    it('handles patch with form parameters', function () {
+
+        this.server.respondWith("PATCH", "/test", function (xhr) {
+            var values = JSON.parse(xhr.requestBody);
+            values.should.have.keys("username","password");
+            values["username"].should.be.equal("joe");
+            values["password"].should.be.equal("123456");
+            var ans = { "passwordok": values["password"] == "123456"};
+            xhr.respond(200, {}, JSON.stringify(ans));
+        });
+
+        var html = make('<form hx-patch="/test" hx-ext="json-enc" > ' + 
+            '<input type="text"  name="username" value="joe"> ' +
+            '<input type="password"  name="password" value="123456"> ' +
+        '<button  id="btnSubmit">Submit</button> ');
+
+        byId("btnSubmit").click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal('{"passwordok":true}');
+    })
+
+    it('handles delete with form parameters', function () {
+
+        this.server.respondWith("DELETE", "/test", function (xhr) {
+            var values = JSON.parse(xhr.requestBody);
+            values.should.have.keys("username","password");
+            values["username"].should.be.equal("joe");
+            values["password"].should.be.equal("123456");
+            var ans = { "passwordok": values["password"] == "123456"};
+            xhr.respond(200, {}, JSON.stringify(ans));
+        });
+
+        var html = make('<form hx-delete="/test" hx-ext="json-enc" > ' + 
+            '<input type="text"  name="username" value="joe"> ' +
+            '<input type="password"  name="password" value="123456"> ' +
+        '<button  id="btnSubmit">Submit</button> ');
+
+        byId("btnSubmit").click();
+        this.server.respond();
+        this.server.lastRequest.response.should.equal('{"passwordok":true}');
+    })
+
+  
+
+});
+
diff --git a/test/index.html b/test/index.html
index 66f6c17..fd667f6 100644
--- a/test/index.html
+++ b/test/index.html
@@ -87,7 +87,7 @@
 <script src="extensions/rails-method.js"></script>
 <script src="extensions/debug.js"></script>
 <script src="extensions/morphdom-swap.js"></script>
-
+<script src="extensions/json-enc.js"></script>
 
 <!-- events last so they don't screw up other tests -->
 <script src="core/events.js"></script>
diff --git a/www/official-extensions.md b/www/official-extensions.md
index 65c3333..c85b589 100644
--- a/www/official-extensions.md
+++ b/www/official-extensions.md
@@ -93,4 +93,29 @@ htmx.defineExtension('morphdom-swap', {
         }
     }
 });
-```
\ No newline at end of file
+```
+
+### <a name="json-enc">[`json-enc`](#json-enc)
+
+#### Description
+
+This extension encodes parameters in JSON format instead of url format.
+
+#### Usage
+
+```html
+<div hx-post='/test' hx-ext='json-enc'>click me</div>
+```
+
+#### Source
+
+```javascript
+htmx.defineExtension('json-enc', {
+    encodeParameters : function(xhr, parameters, elt) {
+        xhr.requestHeaders['Content-Type'] = 'application/json';
+        xhr.overrideMimeType('text/json');
+        return (JSON.stringify(parameters));
+    }
+});        
+
+```

```

chg20 commented:
applied patch here:

https://github.com/bigskysoftware/htmx/commit/3eb4a56e285ec6e0586c10bdaa2a42661f0f3af1

THANK YOU!
